app/src/app/mod.rs

```diff
@@ -32,13 +32,21 @@ pub struct App {
 }
 
 impl App {
-    pub fn new(snapshot: Snapshot) -> Self {
+    pub async fn new(snapshot: Snapshot) -> Result<Self> {
```

Comment:
It feels wrong for this method to be fallible.  Every `Snapshot` should represent a valid chain state, so I think creating an `App`, which is a view of that valid chain state, should also be infallible.

Making it fallible and asynchronous taints all of its callers, who now have to deal with errors, regardless of whether or not that makes sense (e.g., we will have to `await` and `unwrap` every time we reset the mempool fork).

Why don't we just do the check in the one place it matters -- in `commit`?